src/bin/cargo/cli.rs

```diff
@@ -27,6 +27,12 @@ lazy_static::lazy_static! {
 pub fn main(config: &mut LazyConfig) -> CliResult {
     let args = cli().try_get_matches()?;
 
+    // Update the process-level notion of cwd
+    if let Some(new_cwd) = args.get_one::<std::path::PathBuf>("directory") {
+        std::env::set_current_dir(&new_cwd)
+            .with_context(|| "could not change to requested directory")?;
+    }
```

Comment:
Can you revert #11080 and assert that the config is not initialized?  That'll help us catch problems with this in the future

Reply:
Should do the assert outside of the if-check so we don't have to worry as much about coverage for `-C` support